op-service/client/backoff.go

```diff
@@ -0,0 +1,82 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/ethereum-optimism/optimism/op-service/backoff"
+)
+
+var (
+	// ExponentialBackoff is the default backoff strategy.
+	ExponentialBackoff = backoff.Exponential()
+)
+
+// InnerRPC is a minimal [client.RPC] interface that is used by the backoff client.
+//
+//go:generate mockery --name InnerRPC --output ./mocks/
+type InnerRPC interface {
+	Close()
+	CallContext(ctx context.Context, result any, method string, args ...any) error
+	BatchCallContext(ctx context.Context, b []rpc.BatchElem) error
+	EthSubscribe(ctx context.Context, channel any, args ...any) (ethereum.Subscription, error)
+}
+
+// backoffEthClient wraps a [InnerRPC] with a backoff strategy.
+type backoffEthClient struct {
+	c             InnerRPC
+	retryAttempts int
+	strategy      backoff.Strategy
+}
+
+// NewBackoffClient creates a new backoff client.
+// The backoff strategy is optional, if not provided, the default exponential backoff strategy is used.
+func NewBackoffClient(c InnerRPC, retries int, strategy ...backoff.Strategy) *backoffEthClient {
+	if len(strategy) == 0 {
+		strategy = []backoff.Strategy{ExponentialBackoff}
+	}
+	return &backoffEthClient{
```

Comment:
Could this use the instrumented client internally?

Reply:
https://github.com/ethereum-optimism/optimism/blob/develop/op-node/client/rpc.go
Or https://github.com/ethereum-optimism/optimism/blob/develop/op-node/sources/eth_client.go

Reply:
Oh nice! Should I refactor retry.go then into op-service in this pr?

Reply:
It probably is worth refactoring to make it a common thing. Note that op-program is using `maxAttempts = math.MaxInt` because it needs to keep retrying until it works - returning an error will result in a crash.

The max attempts could certainly be passed in as an argument to the retrying source though and then op-program could retry forever and challenger could use something more sensible.

Reply:
> const maxAttempts = math.MaxInt // Succeed or die trying

🔥